File: Collatz.java
public class Collatz {
    public static void collatz(int n) {
        System.out.println(n + " ");
        // base case
        if (n == 1)
            return;
        // recursive case
        else if (n % 2 == 0)
            collatz(n / 2);
        else
            collatz(3 * n + 1);
    }

    public static void main(String[] args) {
        // testcase
        collatz(78);

    }
}

File: Recursion/collatz.py
"""
Visualizing the execution of several 
simple recursive functions
"""


def collatz(num, count):
    """
    Given n, repeatedly perform n = f(n) where
    f(n) = n / 2 if n is even
    f(n) = 3 * n + 1 is n is odd
    Return number of iterations of this redution
    """
    if num == 1:
        # base case
        return count
        # recursive case
    elif (num % 2) == 0:
        return collatz(num / 2, count + 1)
    else:
        return collatz(3 * num + 1, count + 1)


print(collatz(5, 0))
